### PR TITLE
[None][fix] AutoDeploy: Fix shape handling for singleton prefill

### PR DESCRIPTION
**Pull request by galagam**
## Release Notes

* **Bug Fixes**
  * Improved logits handling to correctly preserve tensor shapes for 3D and 2D inputs
  * Enhanced hidden_states normalization during gather operations with better dimension management
  * Added runtime validation for supported input shapes

* **Tests**
  * Added test coverage for 2D tensor format handling in logits operations

## Description

Encountered sampler crashes during development of https://github.com/NVIDIA/TensorRT-LLM/issues/11432 . Possibly related to sampler logic changes in https://github.com/NVIDIA/TensorRT-LLM/pull/11141.
This targeted fix replaces the patchy #11646 draft PR.
The root issue was a shape-contract mismatch in the AD gather path: gather_logits_before_lm_head could receive canonical 2D token-major tensors ([tokens, hidden]), but previously applied squeeze/unsqueeze behavior that implicitly assumed only 3D layouts [b,1,h] or [1,s,h].
Singleton 2D prefill inputs [1,h], collapsed to [h], eventually surfacing as invalid logits shape in sampling. In torch_gather_logits, gathering is now rank-aware: 3D inputs are normalized and restored to their original 3D layout, while 2D inputs are treated as canonical and preserved as 2D. In attention_interface, logits squeezing now operates only on true 3D layout dimensions, preserving canonical 2D logits.

## Test Coverage

tests/unittest/_torch/auto_deploy/unit/singlegpu/transformations/library/test_gather_logits_before_lm_head.py::TestGatherLogitsBeforeLmHeadOp::test_2d_singleton_token_format

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Comment by galagam** [comment]
/bot run --stage-list "DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-1"

File: tensorrt_llm/_torch/auto_deploy/custom_ops/utils/torch_gather_logits.py [hunk 1]
```diff
@@ -25,15 +25,34 @@ def gather_logits_before_lm_head(
...
         Gathered and flattened hidden states [num_gathered_tokens, hidden]
     """
-    # final shape is [total_tokens, hidden] from [b, 1, hidden] or [1, total_tokens, hidden]
-    is_decode_only = hidden_states.shape[1] == 1
-    hidden_states = hidden_states.squeeze(int(is_decode_only))
+    # Normalize to [tokens, hidden] for gather.
+    # NOTE: 2D [tokens, hidden] is already canonical and must not be squeezed on dim 0.
+    if hidden_states.dim() == 3:
+        input_was_3d = True
+        is_decode_only = hidden_states.shape[1] == 1
```
Comment (tcherckez-nvidia) [inline tensorrt_llm/_torch/auto_deploy/custom_ops/utils/torch_gather_logits.py]: maybe we can get along with just one var - `squeezed_dim`? could be [None, 0, 1]

File: tensorrt_llm/_torch/auto_deploy/custom_ops/attention_interface.py [hunk 2]
```diff
@@ -1086,15 +1086,33 @@ def rescatter_input_ids(self, ungathered_input_ids: torch.Tensor):
     @nvtx_range("ad_maybe_gather_logits")
     def maybe_gather_and_squeeze_logits(self, logits: torch.Tensor) -> torch.Tensor:
         """Maybe gather the logits if logits have not been gathered yet."""
-        num_tokens = logits.shape[0] * logits.shape[1]
+        if logits.dim() == 3:
+            num_tokens = logits.shape[0] * logits.shape[1]
```
Comment (tcherckez-nvidia) [inline tensorrt_llm/_torch/auto_deploy/custom_ops/attention_interface.py]: why not initializing it to something outside the if?

**Review by tcherckez-nvidia — APPROVED** [review]

File: tensorrt_llm/_torch/auto_deploy/custom_ops/attention_interface.py (hunk 2 above)
Comment (galagam) [reply tensorrt_llm/_torch/auto_deploy/custom_ops/attention_interface.py]: Do you mean something like this?
I think my version is more readable.
Or something else..?
@tcherckez-nvidia 
```
num_tokens = logits.shape[0] * logits.shape[1]
if logits.dim() == 2:
  num_tokens = logits.shape[0]
if logits.dim() not in (2, 3):
  raise AssertionError(...)
```

Comment (tcherckez-nvidia) [reply tensorrt_llm/_torch/auto_deploy/custom_ops/attention_interface.py]: I don't like defining variables in a inner scope, seems to me like error prone paradigm, if not now in the future when someone will add another branch and will not initialize one of the variables. you can make it a function too.
